### ISSUE TITLE
Ticks and Tick Labels not visible around edge of plot 

### ISSUE THREAD

**Issue by MilkyWayGalaxy3632** [astropy/astropy#16708]
### Description

Hello Astropy Community,

I'm unable to produce ticks along the edges of a plot.

![outputplot2](https://github.com/astropy/astropy/assets/175346433/fc152e74-d0cc-4f88-93f8-88de7e24b534)

Reading the astropy documentation on CoordinateHelper objects https://docs.astropy.org/en/stable/api/astropy.visualization.wcsaxes.CoordinateHelper.html#astropy.visualization.wcsaxes.CoordinateHelper

and applying the following additional parameters including by being very verbose and setting generally default params of the CoordinateHelper object explicitly I am still unable to produce ticks around the edges of the plot. 

```python
glat.set_ticklabel_position('lr')
glat.set_ticklabel_visible(True)
glat.set_ticks_visible(True)
```

I have attempted solutions to similar problems found in
https://github.com/astropy/astropy/issues/7180 , https://github.com/astropy/astropy/pull/9392 , and  https://github.com/astropy/astropy/issues/8521 with no success.

Using interior tickable gridlines, I am able to produce ticks within the plot interior.

```python
glon.add_tickable_gridline('i', -10*u.arcmin)
glat.set_ticks_position('li')
glat.set_ticklabel_position('li')
```
![interiorticksTrue](https://github.com/astropy/astropy/assets/175346433/b888532b-7321-488a-9a8d-95b55c451192)

I have also attempted to position ticks opposite from the axis labels to avoid conflicts between the axis labels and tick labels/ ticks.
```python
glat.set_axislabel_position('l')
glat.set_ticklabel_position('lr')
glat.set_ticks_position('lr')
```

It is worth noting that the axis label will disappear when I set

```python
glon.set_axislabel_visibility_rule('always')
```

to

```python
glon.set_axislabel_visibility_rule('ticks')
```

https://docs.astropy.org/en/stable/api/astropy.visualization.wcsaxes.CoordinateHelper.html#astropy.visualization.wcsaxes.CoordinateHelper.set_axislabel_visibility_rule

I would assume that ticks are not set to visible although I have explicitly set_ticks_visible(True)

### Expected behavior

I would expect to be able to produce plots such as those described in https://docs.astropy.org/en/stable/visualization/wcsaxes/ticks_labels_grid.html

For additional information my wcs object contains,

```
WCS Keywords

Number of WCS axes: 3
CTYPE : 'GLON-CAR' 'GLAT-CAR' 'VRAD' 
CRVAL : 0.0 0.0 0.0 
CRPIX : 3319.0 1660.0 471.921630003202 
PC1_1 PC1_2 PC1_3  : 1.0 0.0 0.0 
PC2_1 PC2_2 PC2_3  : 0.0 1.0 0.0 
PC3_1 PC3_2 PC3_3  : 0.0 0.0 1.0 
CDELT : -0.05427333333333333 0.05427333333333333 1288.21496912415 
NAXIS : 6637  3319  945
```
and fits header
```
SIMPLE  =                    T / conforms to FITS standard                      
BITPIX  =                  -32 / array data type                                
NAXIS   =                    3 / number of array dimensions                     
NAXIS1  =                 6637                                                  
NAXIS2  =                 3319                                                  
NAXIS3  =                  945                                                  
OBJECT  = 'EBHIS   '           / The Effelsberg-Bonn HI Survey                  
TELESCOP= 'Effelsberg 100m RT' / Telescope Name                                 
ORIGIN  = 'AIfA/MPIfR Bonn'    / Organisation or Institution                    
REFERENC= 'Winkel et al. 2016' / A&A 585, A41                                   
RESTFRQ =        1420405751.77                                                  
RESTWAV =       0.211061140541                                                  
CDELT1  = -0.05427333333333333                                                  
CRPIX1  =               3319.0                                                  
CRVAL1  =                  0.0                                                  
CTYPE1  = 'GLON-CAR'                                                            
CUNIT1  = 'deg     '                                                            
CDELT2  =  0.05427333333333333                                                  
CRPIX2  =               1660.0                                                  
CRVAL2  =                  0.0                                                  
CTYPE2  = 'GLAT-CAR'                                                            
CUNIT2  = 'deg     '                                                            
WCSAXES =                    3                                                  
LONPOLE =                  0.0                                                  
LATPOLE =                 90.0                                                  
CRVAL3  =                  0.0                                                  
CRPIX3  =     471.921630003202                                                  
CDELT3  =     1288.21496912415                                                  
CUNIT3  = 'm/s     '                                                            
CTYPE3  = 'VRAD    '                                                            
CNAME3  = 'LSRK radio velocity'                                                 
SPECSYS = 'LSRK    '                                                            
SSYSOBS = 'LSRK    '                                                            
SSYSSRC = 'LSRK    '                                                            
VELOSYS =                  0.0                                                  
ZSOURCE =                  0.0                                                  
BUNIT   = 'K       '                                                            
BPA     =                  0.0                                                  
BMAJ    =          0.180317698                                                  
BMIN    =          0.180317698                                                  
EQUINOX =               2000.0                                                  
CHECKSUM= 'aQGmaODjaODjaODj'   / HDU checksum updated 2023-01-09T13:39:30       
DATASUM = '810657741'          / data unit checksum updated 2023-01-09T13:39:30 
COMMENT Please also cite Kerp et al. 2011, AN 332, 637                               
```

### How to Reproduce

```python
import matplotlib.pyplot as plt
import matplotlib as mpl
from astropy.wcs import WCS
from astropy.io import fits
from astropy import units as u

# Load the FITS file
EBHIS = 'EBHIS_GAL_CAR_All_Sky_HI_Survey-001.fits'

data_ebhis_2 = fits.getdata(EBHIS)
hdr_ebhis_2 = fits.getheader(EBHIS)
wcs_ebhis_2 = WCS(hdr_ebhis_2)

# Create the plot
fig = plt.figure(figsize=(10, 8))
ax = fig.add_subplot(111, projection=wcs_ebhis_2, slices=('x', 'y', 472))

im = ax.imshow(data_ebhis_2[472, :, :], cmap='viridis', origin='lower')
ax.coords.grid(color='white', alpha=0.5, linestyle='solid')

# Galactic Longitude settings
glon = ax.coords['glon']
glon.set_axislabel('Galactic Longitude')
glon.set_axislabel_position('b')
glon.set_ticklabel_position('tb')
glon.set_ticks_position('tb')
glon.set_axislabel_visibility_rule('always')
glon.set_ticklabel_visible(True)
glon.set_ticks_visible(True)
glon.set_ticks(values=[0, 45, 90, 135, 180, 225, 270, 315, 360] * u.degree)
glon.tick_params(width=2, length=6, direction='in', labelsize=10, colors='red')

# Galactic Latitude settings
glat = ax.coords['glat']
glat.set_axislabel('Galactic Latitude')
glat.set_axislabel_position('l')
glat.set_ticklabel_position('lr')
glat.set_ticks_position('lr')
glat.set_axislabel_visibility_rule('always')
glat.set_ticklabel_visible(True)
glat.set_ticks_visible(True)
glat.set_ticks(values=[-90, -60, -30, 0, 30, 60, 90] * u.degree)
glat.tick_params(width=2, length=6, direction='in', labelsize=10, colors='blue')

plt.colorbar(im, ax=ax, orientation='vertical', pad=0.05, aspect=50, shrink = 0.5)
plt.show()

```
![outputplot](https://github.com/astropy/astropy/assets/175346433/211f8214-0c21-4e9e-9e78-8445d2d374f7)

### Versions

```
Linux-4.4.0-19041-Microsoft-x86_64-with-glibc2.35
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
astropy 6.1.1
Numpy 1.26.0
pyerfa 2.0.1.4
Scipy 1.14.0
Matplotlib 3.5.1
```

**Comment by ayshih** [comment]
(Incidentally, I think it's awesome that you tested using tickable gridlines, because I wrote that!)

This is essentially a duplicate of #10201 and #9784, and has nothing to do with visibility settings.  The problem is that your data array (6637 pixels by 3319 pixels) is larger than the "physical" coordinate ranges for longitude and latitude.  One way to see this (note the NaNs):
```python
>>> wcs_ebhis_2.pixel_to_world_values(0, 0, 0)
(array(nan), array(nan), array(-606648.29305447))
>>> wcs_ebhis_2.pixel_to_world_values(6636, 3318, 0)
(array(nan), array(nan), array(-606648.29305447))
```
To see what pixel indices are actually physical:
```python
>>> wcs_ebhis_2.world_to_pixel_values(180 - 1e-10, -90, 0)
(array(1.45412112), array(0.72706056), array(470.92163))
>>> wcs_ebhis_2.world_to_pixel_values(180 + 1e-10, +90, 0)
(array(6634.54587888), array(3317.27293944), array(470.92163))
```
That is, longitude runs from pixel 1.454 to pixel 6634.546 (counting from zero) and latitude runs from pixel 0.727 to pixel 3317.273 (again, counting from zero).  When Astropy WCSAxes plots the full extent of your data array (6637 pixels by 3319 pixels), the default pixel ranges are -0.5 to 6636.5 in longitude and -0.5 to 3318.5 in latitude.  What this means for your data array is that the boundary edges of the plot at default limits do not correspond to physical coordinates, so WCSAxes concludes that the edges have no ticks.

What you need to do is to crop the plot limits to be within physical coordinates, e.g.:
```python
ax.set_xlim(1.5, 6634.5)
ax.set_ylim(1, 3317)
```
With this change, the ticks along the left and bottom edges should show up and have labels.

For some reason I don't understand, potentially a WCSAxes bug, the above doesn't fix the ticks and labels along the right and top edges.  I find that I have to contract the plots limits far more than should be needed to get those edges to work, e.g.:
```python
ax.set_xlim(1.5, 6617)
ax.set_ylim(1, 3300)
```
That potential bug warrants further investigation, but the change needed for the left and bottom edges makes sense.

**Comment by ayshih** [comment]
To describe the problem conceptually, WCSAxes puts a tick wherever a gridline crosses a boundary edge.  When the plot limits exceed physical coordinates, the gridlines never actually touch a boundary edge – they just stop in the middle of the plot – and thus there are no ticks.

**Comment by ayshih** [comment]
> For some reason I don't understand, potentially a WCSAxes bug, the above doesn't fix the ticks and labels along the right and top edges.

It took me two years to look into this and fix this bug (#19057), but it's finally fixed!

**Comment by pllim** [comment]
Two years is a blink of an eye for astronomers, right, right? You have like, what, a few million more years before a photon makes its way onto Sun's surface?
